File: leetcode/valid-parenthese.ts
function isValid(s: string): boolean {
  const stack: string[] = [];
  const map: Record<string, string> = {
    "(": ")",
    "{": "}",
    "[": "]",
  };

  for (const char of s) {
    if (char === map[stack[stack.length - 1]]) {
      stack.pop();
    } else {
      stack.push(char);
    }
  }
  return stack.length === 0;
}

File: leetcode/valid-parenthese.go
package leetcode

func isValid(s string) bool {
	stack := []rune{}
	bracketMap := map[rune]rune{
		')': '(',
		'}': '{',
		']': '[',
	}
	for _, char := range s {
		if _char, ok := bracketMap[char]; ok && len(stack) > 0 && stack[len(stack)-1] == _char {
			stack = stack[:len(stack)-1]
		} else {
			stack = append(stack, char)
		}
	}

	return len(stack) == 0
}
